persistent-postgresql/Database/Persist/Postgresql/Util.hs

```diff
@@ -0,0 +1,30 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Database.Persist.Postgresql.Util (
+  mkPostgresUpdateText
+) where
+
+import Data.Monoid ((<>))
+import Data.Text (Text)
+import qualified Data.Text as T
+
+import Database.Persist (FieldNameDB, PersistEntity (..), PersistUpdate(..), Update(..), updateUpdate, fieldDB, FieldDef)
+
+
+mkPostgresUpdateText :: PersistEntity record => (FieldNameDB -> Text) -> (Text -> Text) -> Update record -> Text
+mkPostgresUpdateText escapeName refColumn x =
+  case updateUpdate x of
+    Assign -> n <> "=?"
+    Add -> T.concat [n, "=EXCLUDED.", refColumn n, "+?"]
+    Subtract -> T.concat [n, "=EXCLUDED.", refColumn n, "-?"]
+    Multiply -> T.concat [n, "=EXCLUDED.", refColumn n, "*?"]
+    Divide -> T.concat [n, "=EXCLUDED.", refColumn n, "/?"]
+    BackendSpecificUpdate up ->
+      error . T.unpack $ "mkUpdateText: BackendSpecificUpdate " <> up <> " not supported"
+  where
+    n = escapeName . fieldDB . updateFieldDef $ x
+
+-- | Gets the 'FieldDef' for an 'Update'.
+updateFieldDef :: PersistEntity v => Update v -> FieldDef
+updateFieldDef (Update f _ _) = persistFieldDef f
+updateFieldDef BackendUpdate {} = error "updateFieldDef: did not expect BackendUpdate"
```

Comment:
For a general purpose library function, we shouldn't ever call `error` when we're handling inputs. Instead I'd write this with a `Maybe FieldDef` return type.

Also, since this is only used in `mkPostgresUpdateText`, we can inline it directly and perform the pattern match there.